### PR TITLE
Performance/resolve page tree

### PR DESCRIPTION
**Pull request by motechFR**
### WHAT
copilot:summary

### WHY
Improve performance of resolve page tree up to 20x on the raw query, and bring mapTree operation from 250ms to 0.

Tested locally on a space with 8 000 pages

This uses a recursive raw SQL query to handle tree resolution, so our code won't need to load all the space pages anymore when resolving page tree for a specific page.

### CODE REVIEW

File: src/lib/pages/resolvePageTree.ts [hunk 1]
```diff
@@ -108,35 +114,55 @@ export async function resolvePageTree({
...
+    UNION ALL
+    
+    SELECT p.id, p."parentId"
+    FROM "Page" p
+    INNER JOIN children_cte ON p."parentId" = children_cte.id
+  )
+  SELECT id FROM parents_cte 
+  UNION 
+  SELECT id FROM children_cte;
+`) as Pick<Page, 'id'>[];
```
Comment (mattcasey) [inline src/lib/pages/resolvePageTree.ts:157]: Neat.. are there any gotchas using Recursive queries? can it go on forever  if there is a circular dependency?

File: src/lib/pages/resolvePageTree.ts [hunk 2]
```diff
@@ -161,7 +187,7 @@ export async function resolvePageTree({
     parents,
     targetPage,
     flatChildren: flattenChildren ? flattenTree(targetPage) : undefined
-  };
+  } as any;
```
Comment (mattcasey) [inline src/lib/pages/resolvePageTree.ts:190]: this `any` doesn't seem necessary - tsc compiles fine for me

**Review by mattcasey — APPROVED** [review]
Really exciting! Thanks for working this out. I'm assuming we already have tests + all permissions tests should be adding extra coverage

File: src/lib/pages/resolvePageTree.ts (hunk 1 above)
Comment (motechFR) [reply src/lib/pages/resolvePageTree.ts]: Thanks alot for raising this! I added some safeguards in the SQL query, and the reducePagesToPageTree to handle this

File: src/lib/pages/resolvePageTree.ts (hunk 2 above)
Comment (motechFR) [reply src/lib/pages/resolvePageTree.ts]: Cleaned up :)

**Comment by motechFR** [comment]
@mattcasey Yep, we already had great test coverage on all code paths relating to the code modified here, so I expect the internal implementation switch to be safe

**Pull request merged**